proposals/NNNN-Promises-in-Policy-Server.md

```diff
@@ -0,0 +1,114 @@
+
+## Motivation
+
+The last structural change to the policy server was written at a time where Promises were not as well understood, and so the established method of handling asynchronous tasks was using callbacks. Earlier attempts to mitigate the issues of readability with this method have not scaled so well with the number of features the project has accumulated. The codebase has grown to a point where any additions or changes to the application logic require a significant amount of reading and care to ensure that no new bugs are introduced and that all cases are considered.
+
+With the introduction of await/async syntax, asynchronous logic can be read from top to bottom without excessive nesting of functions or visually jumping around the method to keep track of the execution order. This Promise style syntax will require fewer lines and will automatically throw functions into the event loop instead of being ran synchronously.
+
+## Proposed solution
+
+All callback-based asynchronous management that resides in the `app/v1` directory will be changed to use Promises and async/await syntax. If using a function from a library that uses callback-based, error-first style and is outside the project's control, it will be converted into a Promise through the use of NodeJS's `util.promisify` helper method. All references to the project's flame, flow, and async libraries to help with asynchronous function chaining will be replaced entirely with Promises and async/await syntax. Express middleware functions will necessarily remain using the `next` callback style, but if possible this will be avoided elsewhere.
+
+All controller, helper, and model files are suspectible to these changes. The sql files will be untouched as they merely construct SQL statements. The modules inside the `custom` folder such as the cache and database modules should have their methods change to return Promises. Due to the expectancy of developers to change the contents inside the `customizable` folder with their own code, no changes will be required for those exported functions. The only requirement to use this new feature should be to upgrade the NodeJS version. The server can convert these exported functions that use callbacks into ones that return Promises anyway.
+
+The front-end's use of callbacks are much more controlled: the rare use of asynchronous function chaining means the code is still readable. Changing some of the common functions used such as the Vue mixin `httpRequest` would not have significant benefits because of the necessary two-value return of an error and a result. Catching errors in a promise still require a callback to be passed in, which leads to the same problem as callback-based methods if you are not chaining multiple asynchronous functions together. For these reasons, the front-end will not be changed for this proposal.
```

Comment:
```suggestion
The front-end's use of callbacks is much more controlled: the rare use of asynchronous function chaining means the code is still readable. Changing some of the common functions used such as the Vue mixin `httpRequest` would not have significant benefits because of the necessary two-value return of an error and a result. Catching errors in a Promise still require a callback to be passed in, which leads to the same problem as callback-based methods if you are not chaining multiple asynchronous functions together. For these reasons, the front-end will not be changed for this proposal.
```